File: src/main/java/io/mfedirko/common/infra/dynamodb/DynamoDbContactMeRepository.java
package io.mfedirko.common.infra.dynamodb;

import io.mfedirko.common.util.DateHelper;
import io.mfedirko.contactme.ContactForm;
import io.mfedirko.contactme.ContactHistory;
import io.mfedirko.contactme.ContactMeRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.context.annotation.Profile;
import org.springframework.stereotype.Repository;
import software.amazon.awssdk.enhanced.dynamodb.*;
import software.amazon.awssdk.enhanced.dynamodb.model.PageIterable;

import java.time.*;
import java.util.Collections;
import java.util.List;
import java.util.Objects;

import static io.mfedirko.common.util.DateHelper.TZ_UTC;
import static software.amazon.awssdk.enhanced.dynamodb.model.QueryConditional.sortBetween;
import static software.amazon.awssdk.enhanced.dynamodb.model.QueryConditional.sortGreaterThanOrEqualTo;

@Repository
@RequiredArgsConstructor
@Profile("!mock")
@Slf4j
public class DynamoDbContactMeRepository implements ContactMeRepository {
    private final DynamoDbEnhancedClient enhancedClient;

    @Override
    public void save(ContactForm form) {
        Objects.requireNonNull(form, "ContactForm must be non-null");

        getTable().putItem(DynamoContactRequest.from(form));
    }

    @Override
    public List<ContactHistory> findContactHistoryByDate(LocalDate date) {
        Objects.requireNonNull(date, "Date must be non-null");

        PageIterable<DynamoContactRequest> result = getTable().query(k -> k.scanIndexForward(false)
                .queryConditional(sortBetween(
                        toKey(DateHelper.toUtcStartOfDay(date)),
                        toKey(DateHelper.toUtcEndOfDay(date))))
                .build());
        if (result == null) {
            log.warn("DynamoDB returned null for date {}", date);
            return Collections.emptyList();
        }
        return result.items().stream()
                .map(DynamoContactRequest::toContactHistory)
                .toList();
    }
    private DynamoDbTable<DynamoContactRequest> getTable() {
        return enhancedClient.table(DynamoContactRequest.TABLE, TableSchema.fromBean(DynamoContactRequest.class));
    }

    private static Key toKey(LocalDateTime date) {
        return Key.builder()
                .partitionValue(DynamoContactRequest.toPartitionKey(date.toLocalDate()))
                .sortValue(toSortKey(date))
                .build();
    }

    private static long toSortKey(LocalDateTime date) {
        Instant instant = date.atZone(TZ_UTC).toInstant();
        return DynamoContactRequest.toSortKey(instant);
    }
}

File: src/main/kotlin/io/mfedirko/common/infra/dynamodb/DynamoDbContactMeRepository.kt
package io.mfedirko.common.infra.dynamodb

import io.mfedirko.common.util.DateHelper.TZ_UTC
import io.mfedirko.common.util.DateHelper.toUtcEndOfDay
import io.mfedirko.common.util.DateHelper.toUtcStartOfDay
import io.mfedirko.common.util.logger
import io.mfedirko.contactme.ContactForm
import io.mfedirko.contactme.ContactHistory
import io.mfedirko.contactme.ContactMeRepository
import org.springframework.context.annotation.Profile
import org.springframework.stereotype.Repository
import software.amazon.awssdk.enhanced.dynamodb.DynamoDbEnhancedClient
import software.amazon.awssdk.enhanced.dynamodb.DynamoDbTable
import software.amazon.awssdk.enhanced.dynamodb.Key
import software.amazon.awssdk.enhanced.dynamodb.TableSchema
import software.amazon.awssdk.enhanced.dynamodb.model.PageIterable
import software.amazon.awssdk.enhanced.dynamodb.model.QueryConditional
import software.amazon.awssdk.enhanced.dynamodb.model.QueryEnhancedRequest
import java.time.LocalDate
import java.time.LocalDateTime
import java.util.*

@Repository
@Profile("!mock")
class DynamoDbContactMeRepository(
    private val enhancedClient: DynamoDbEnhancedClient

) : ContactMeRepository {
    private val log = logger()

    private val table: DynamoDbTable<DynamoContactRequest>
        get() = enhancedClient.table(
            DynamoContactRequest.TABLE, TableSchema.fromBean(
                DynamoContactRequest::class.java
            )
        )

    override fun save(form: ContactForm) {
        table.putItem(DynamoContactRequest.from(form))
    }

    override fun findContactHistoryByDate(date: LocalDate): List<ContactHistory> {
        val result: PageIterable<DynamoContactRequest>? = table.query { query: QueryEnhancedRequest.Builder ->
            query.scanIndexForward(false)
                .queryConditional(
                    QueryConditional.sortBetween(
                        toKey(date.toUtcStartOfDay()),
                        toKey(date.toUtcEndOfDay())
                    )
                )
                .build()
        }

        return result?.items()
            ?.map { it.toContactHistory() }
            ?.toList()
            ?: emptyList()
    }


    companion object {
        private fun toKey(date: LocalDateTime): Key {
            return Key.builder()
                .partitionValue(DynamoContactRequest.toPartitionKey(date.toLocalDate()))
                .sortValue(toSortKey(date))
                .build()
        }

        private fun toSortKey(date: LocalDateTime): Long {
            val instant = date.atZone(TZ_UTC).toInstant()
            return DynamoContactRequest.toSortKey(instant)
        }
    }
}